### PR TITLE
[Transformer] Simplify transformer reader and fix TranslationDataset

### PR DESCRIPTION
**Pull request by FrostML**
* Fix translation dataset. 
* Provide delimiter for tokenizer. 
* Add WMT14 dataset. 
* Simplify transformer reader. 
* Delete gen_data.sh 

ready. plz review this PR.

### CODE REVIEW

File: PaddleNLP/examples/machine_translation/transformer/reader.py [hunk 1]
```diff
@@ -22,67 +22,50 @@
...
-            sort_type=args.sort_type,
-            shuffle=args.shuffle,
-            shuffle_batch=args.shuffle_batch,
-            use_token_batch=args.use_token_batch,
-            max_length=args.max_length,
-            distribute_mode=True if i == 0 else False,
-            world_size=world_size,
-            rank=rank)
+            drop_last=False,
+            batch_size_fn=token_size_fn).shard()
```
Comment (guoshengCS) [inline PaddleNLP/examples/machine_translation/transformer/reader.py]: 可以再确认下，按照这里的设计，要实现之前token组batch的功能，将`batch_size_fn`返回的size_so_far表示当前minibatch中已有样本的最大长度，`key`返回`max_length * minibatch_len`即可。可以参考之前的PR https://github.com/PaddlePaddle/models/pull/4991#discussion_r538043679

Comment (guoshengCS) [reply PaddleNLP/examples/machine_translation/transformer/reader.py]: 类似这里的例子 https://github.com/PaddlePaddle/models/blob/develop/PaddleNLP/paddlenlp/datasets/translation.py#L296 ，可以确认下

Comment (FrostML) [reply PaddleNLP/examples/machine_translation/transformer/reader.py]: 好的。之前是想要将 so far 作为 sum of tokens，key 直接用默认，已经修改。谢谢。

File: PaddleNLP/paddlenlp/datasets/translation.py [hunk 2]
```diff
@@ -273,6 +273,66 @@ def __init__(self, mode='train', root=None, transform_func=None):
                           transform_func[1](data[1])) for data in self.data]
 
 
+class WMT14ende(TranslationDataset):
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/datasets/translation.py:276]: 因为使用了BPE的数据，是否要对此进行说明

File: PaddleNLP/paddlenlp/datasets/translation.py [hunk 3]
```diff
@@ -273,6 +273,66 @@ def __init__(self, mode='train', root=None, transform_func=None):
...
+    Args:
+        mode(str, optional): It could be 'train', 'dev' or 'test'. Default: 'train'.
+        root(str, optional): If None, dataset will be downloaded in
+            `/root/.paddlenlp/datasets/machine_translation/WMT14ende/`. Default: None.
+        transform_func(callable, optional): If not None, it transforms raw data
+            to index data. Default: None.
+    Examples:
+        .. code-block:: python
+            from paddlenlp.datasets import WMT14ende
+            train_dataset = WMT14ende.get_datasets(mode="train", transform_func=transform_func)
```
Comment (guoshengCS) [inline PaddleNLP/paddlenlp/datasets/translation.py:290]: 实例代码中也加上`transform_func` 的创建保证跑通吧

File: PaddleNLP/paddlenlp/datasets/translation.py (hunk 2 above)
Comment (guoshengCS) [reply PaddleNLP/paddlenlp/datasets/translation.py:276]: 另外如果name中使用WMT14的话最好还是确认下使用的数据是否不包含WMT14以后的patch吧

File: PaddleNLP/examples/machine_translation/transformer/README.md [hunk 4]
```diff
@@ -91,9 +99,9 @@ python predict.py
 # 还原 predict.txt 中的预测结果为 tokenize 后的数据
 sed -r 's/(@@ )|(@@ ?$)//g' predict.txt > predict.tok.txt
 # 若无 BLEU 评估工具，需先进行下载
-# git clone https://github.com/moses-smt/mosesdecoder.git
+git clone https://github.com/moses-smt/mosesdecoder.git
 # 以英德翻译 newstest2014 测试数据为例
-perl gen_data/mosesdecoder/scripts/generic/multi-bleu.perl gen_data/wmt14_ende_data/newstest2014.tok.de < predict.tok.txt
+perl mosesdecoder/scripts/generic/multi-bleu.perl ~/.paddlenlp/datasets/machine_translation/WMT14ende/WMT14.en-de/wmt14_ende_data/newstest2014.tok.de < predict.tok.txt
```
Comment (guoshengCS) [inline PaddleNLP/examples/machine_translation/transformer/README.md:104]: 后面也可以看下是否可以接`paddlenlp.metric`中的BLEU使用，可以看下能否在predict.py中加上用`paddlenlp.metric.BLEU`评估并打印结果，这里稍多一步的是需要转成非BPE的数据

File: PaddleNLP/paddlenlp/datasets/translation.py [hunk 5]
```diff
@@ -29,8 +29,8 @@ def get_default_tokenizer():
     """Only support split tokenizer
     """
 
-    def _split_tokenizer(x):
-        return x.split()
+    def _split_tokenizer(x, delimiter=None):
```
Comment (LiuChiachi) [inline PaddleNLP/paddlenlp/datasets/translation.py:32]: 这里需要加`delimeter`这个参数吗

Comment (FrostML) [reply PaddleNLP/paddlenlp/datasets/translation.py:32]: 我理解，可以增加一个用于切分的选项，可用于外部使用或是升级 `get_default_transform_func`。

File: PaddleNLP/benchmark/transformer/dygraph/train.py [hunk 6]
```diff
@@ -51,9 +51,8 @@ def do_train(args):
         paddle.seed(random_seed)
 
     # Define data loader
-    (train_loader, train_steps_fn), (eval_loader,
-                                     eval_steps_fn) = reader.create_data_loader(
-                                         args, trainer_count, rank)
+    (train_loader, train_steps_fn), (
+        eval_loader, eval_steps_fn) = reader.create_data_loader(args)
```
Comment (guoshengCS) [inline PaddleNLP/benchmark/transformer/dygraph/train.py]: 如果没有使用的话这里要不先把train_steps_fn和eval_steps_fn给去掉吧，之前加上应该是因为使用高级API代码中定制化的callback中使用到了 https://github.com/PaddlePaddle/hapi/blob/master/transformer/train.py#L35 。

File: PaddleNLP/benchmark/transformer/reader.py [hunk 7]
```diff
@@ -22,67 +22,58 @@
...
-            end_mark=args.special_token[1],
-            unk_mark=args.special_token[2])
-        args.src_vocab_size, args.trg_vocab_size, args.bos_idx, args.eos_idx, \
-            args.unk_idx = dataset.get_vocab_summary()
-        batch_sampler = TransformerBatchSampler(
-            dataset=dataset,
+    for i, dataset in enumerate(datasets):
+        sampler = SamplerHelper(
+            dataset.filter(partial(
+                min_max_filer, max_len=args.max_length)))
```
Comment (guoshengCS) [inline PaddleNLP/benchmark/transformer/reader.py]: 由于dataset.filter返回一个新的dataset而非修改原来的dataset，这里只是将SamplerHelper使用的dataset换成过滤长度的dataset是否会有潜在的问题，此时SamplerHelper与DataLoader使用的dataset不一致，dataloader使用SamplerHelper产生的index取dataset中的数据时还是可能拿到超长的数据。

File: PaddleNLP/examples/machine_translation/transformer/README.md [hunk 8]
```diff
@@ -46,6 +45,15 @@
...
 
+同时，我们提供了一份已经处理好的数据集，可以编写如下代码，对应的数据集将会自动下载并且解压到 `~/.paddlenlp/datasets/machine_translation/WMT14ende/`。
+
+``` python
+# 获取默认的数据处理方式
+transform_func = WMT14ende.get_default_transform_func(root=root)
+# 下载并处理 WMT14.en-de 翻译数据集
+dataset = WMT14ende.get_datasets(mode="train", transform_func=transform_func)
+```
+
```
Comment (guoshengCS) [inline PaddleNLP/examples/machine_translation/transformer/README.md:56]: 后续可以看下加下如何使用自定义数据，可以简单说明，毕竟我们期望的是使用paddlenlp的api后能够更容易的接入数据，原来的README中是有简单说明的，不能说使用paddlenlp的api后反倒不支持自定义数据了。

File: PaddleNLP/benchmark/transformer/reader.py (hunk 7 above)
Comment (guoshengCS) [reply PaddleNLP/benchmark/transformer/reader.py]: 可以将dataset替换为dataset.filter的返回值

File: PaddleNLP/paddlenlp/datasets/translation.py (hunk 2 above)
Comment (FrostML) [reply PaddleNLP/paddlenlp/datasets/translation.py:276]: 因数据集对齐，涉及一条 wmt17 的数据，参考竞品仍使用 wmt14 作为名称。

File: PaddleNLP/paddlenlp/datasets/translation.py (hunk 3 above)
Comment (FrostML) [reply PaddleNLP/paddlenlp/datasets/translation.py:290]: Done. Thanks.

File: PaddleNLP/benchmark/transformer/dygraph/train.py (hunk 6 above)
Comment (FrostML) [reply PaddleNLP/benchmark/transformer/dygraph/train.py]: Done. Thanks.

File: PaddleNLP/benchmark/transformer/reader.py (hunk 7 above)
Comment (FrostML) [reply PaddleNLP/benchmark/transformer/reader.py]: Done. Thanks.

File: PaddleNLP/examples/machine_translation/transformer/README.md (hunk 4 above)
Comment (FrostML) [reply PaddleNLP/examples/machine_translation/transformer/README.md:104]: 结果有误，后面需调试后修改。

**Review by guoshengCS — APPROVED** [review]

File: PaddleNLP/benchmark/transformer/reader.py [hunk 9]
```diff
@@ -22,67 +22,67 @@
...
+        if args.sort_type == SortType.GLOBAL:
+            buffer_size = -1
+            trg_key = (lambda x, data_source: len(data_source[x][1]) + 1)
+            # Sort twice
+            sampler = sampler.sort(
+                key=trg_key, buffer_size=buffer_size).sort(
+                    key=src_key, buffer_size=buffer_size)
+        else:
+            sampler = sampler.shuffle()
+            if args.sort_type == SortType.POOL:
```
Comment (guoshengCS) [inline PaddleNLP/benchmark/transformer/reader.py:73]: 刚才追查hung住的问题发现eval数据用了shuffle，配上token组batch导致各卡各进程eval数据的batch数不一致。后来想了下应该不止是这个，同样的原因，train的数据可能也会多进程多卡不一致，训练时每个batch还必须同步，这个大概率会导致hung住。这里用global sort或者将shuffle设置下seed吧，这样应该能够避免。 还请切记这里的坑 @FrostML
